File: assignments/submissions/40230259/01/python/matrizes_tipos.py
import numpy as np

# Task 1
print("\nTarefa 1")

# Matriz de zeros (3x4)
zero_matrix = np.zeros((3, 4))

# Matriz identidade (4x4)
identity_matrix = np.eye(4)

# Matriz diagonal (3x3) com os valores 2, 5 e -1
diagonal_matrix = np.diag([2, 5, -1])

# Matriz triangular superior (3x3)
upper_triangular = np.array([
    [3, 1, 6],
    [0, 2, 9],
    [0, 0, 9]
], dtype=float)

# Matriz triangular inferior (3x3)
lower_triangular = np.array([
    [3, 0, 0],
    [6, 2, 0],
    [7, 7, 9]
], dtype=float)


# Matriz simétrica, a partir da fórmula M + M^T
M = np.array ([
    [3, 1, 6],    
    [6, 2, 9],
    [7, 7, 9]
], dtype=float)

symmetric_matrix = M + M.T


print("Matriz de Zeros:")
print(zero_matrix)

print("\nMatriz Identidade:")
print(identity_matrix)

print("\nMatriz Diagonal:")
print(diagonal_matrix)

print("\nMatriz Triangular Superior:")
print(upper_triangular)

print("\nMatriz Triangular Inferior:")
print(lower_triangular)

print("\nMatriz Simétrica (M + M^T):")
print(symmetric_matrix)



# Task 2
print("\nTarefa 2")

print("\nMatriz de Zeros:")
print(f"Dimensões: {zero_matrix.shape[0]} x  {zero_matrix.shape[1]}")
print(f"a23 = {zero_matrix[1, 2]}")

print("\nMatriz Identidade:")
print(f"Dimensões: {identity_matrix.shape[0]} x  {identity_matrix.shape[1]}")
print(f"a23 = {identity_matrix[1, 2]}")

print("\nMatriz Diagonal:")
print(f"Dimensões: {diagonal_matrix.shape[0]} x  {diagonal_matrix.shape[1]}")
print(f"a23 = {diagonal_matrix[1, 2]}")

print("\nMatriz Triangular Superior:")
print(f"Dimensões: {upper_triangular.shape[0]} x  {upper_triangular.shape[1]}")
print(f"a23 = {upper_triangular[1, 2]}")

print("\nMatriz Triangular Inferior:")
print(f"Dimensões: {lower_triangular.shape[0]} x  {lower_triangular.shape[1]}")
print(f"a23 = {lower_triangular[1, 2]}")

print("\nMatriz Simétrica:")
print(f"Dimensões: {symmetric_matrix.shape[0]} x  {symmetric_matrix.shape[1]}")
print(f"a23 = {symmetric_matrix[1, 2]}")



# Task 3
print("\nTarefa 3")

def classificar_matriz(A):
    classificacoes = []
    linhas, colunas = A.shape

    # Quadrada e Retangular
    if linhas == colunas:
        classificacoes.append("quadrada")
    else:
        classificacoes.append("retangular")

    # Nula
    if np.all(A == 0):
        classificacoes.append("nula")

    # Identidade
    if linhas == colunas and np.allclose(A, np.eye(linhas)):
        classificacoes.append("identidade")

    # Diagonal
    if linhas == colunas and np.allclose(A, np.diag(np.diag(A))):
        classificacoes.append("diagonal")

    # Triangular Superior
    if linhas == colunas and np.allclose(A, np.triu(A)):
        classificacoes.append("triangular_superior")    

    # Triangular Inferior
    if linhas == colunas and np.allclose(A, np.tril(A)):
        classificacoes.append("triangular_inferior")    

    return classificacoes

print(f"Matriz de Zeros: {classificar_matriz(zero_matrix)}")

print(f"Matriz Identidade: {classificar_matriz(identity_matrix)}")

print(f"Matriz Diagonal: {classificar_matriz(diagonal_matrix)}")

print(f"Matriz Triangular Superior: {classificar_matriz(upper_triangular)}")

print(f"Matriz Triangular Inferior: {classificar_matriz(lower_triangular)}")

print(f"Matriz Simétrica (M + M^T): {classificar_matriz(symmetric_matrix)}")



# Task 4
print("\nTarefa 4")

# Soma de matrizes incompatíveis
A = np.ones((6, 2))
B = np.ones((3, 5))

try:
    resultado = A + B
except ValueError as erro:
    print("Erro ao somar matrizes incompatíveis:")
    print(erro)

# Multiplicação de matrizes incompatíveis
    C = np.ones((2, 7))
    D = np.ones((4, 3))

try:
    resultado = C @ D
except ValueError as erro:
    print("\nErro ao multiplicar matrizes incompatíveis:")
    print(erro)

File: assignments/submissions/40250178/01/Javascript/matrizes_tipos.js
const math = require('mathjs');

// ─── T1: Criar matrizes especiais ─────────────────────────────────────────────

const zeroMatrix = math.zeros(3, 4);
console.log('Matriz Zero (3x4):');
console.table(zeroMatrix.toArray());

const identityMatrix = math.identity(4);
console.log('Matriz Identidade (4x4):');
console.table(identityMatrix.toArray());

const diagonalMatrix = math.diag([2, 5, -1]);
console.log('Matriz Diagonal [2, 5, -1]:');
console.table(diagonalMatrix.toArray());

const upperTriangular = [
  [1, 2, 3],
  [0, 4, 5],
  [0, 0, 6]
];
console.log('Matriz Triangular Superior (3x3):');
console.table(upperTriangular);

const lowerTriangular = [
  [1, 0, 0],
  [2, 3, 0],
  [4, 5, 6]
];
console.log('Matriz Triangular Inferior (3x3):');
console.table(lowerTriangular);

const M = math.matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]]);
const symmetricMatrix = math.add(M, math.transpose(M));
console.log('Matriz Simétrica (M + M^T):');
console.table(symmetricMatrix.toArray());

// ─── T2: Dimensões e elemento a[2][3] ─────────────────────────────────────────

function printInfo(name, matrix) {
  const arr = Array.isArray(matrix) ? matrix : matrix.toArray();
  const rows = arr.length;
  const cols = arr[0].length;
  const val = (rows > 2 && cols > 3) ? arr[2][3] : 'N/A (fora dos limites)';
  console.log(`\n${name}: ${rows}x${cols}, a[2][3] = ${val}`);
}

printInfo('Zero 3x4', zeroMatrix);
printInfo('Identidade 4x4', identityMatrix);
printInfo('Diagonal 3x3', diagonalMatrix);
printInfo('Triangular Superior 3x3', upperTriangular);
printInfo('Triangular Inferior 3x3', lowerTriangular);
printInfo('Simétrica 3x3', symmetricMatrix);

// ─── T3: Função classificar_matriz ────────────────────────────────────────────

function isSquare(arr) {
  return arr.every(row => row.length === arr.length);
}

function isZero(arr) {
  return arr.every(row => row.every(v => Math.abs(v) < 1e-10));
}

function isIdentity(arr) {
  if (!isSquare(arr)) return false;
  const n = arr.length;
  for (let i = 0; i < n; i++)
    for (let j = 0; j < n; j++)
      if (Math.abs(arr[i][j] - (i === j ? 1 : 0)) > 1e-10) return false;
  return true;
}

function isDiagonal(arr) {
  if (!isSquare(arr)) return false;
  for (let i = 0; i < arr.length; i++)
    for (let j = 0; j < arr[i].length; j++)
      if (i !== j && Math.abs(arr[i][j]) > 1e-10) return false;
  return true;
}

function isSymmetric(arr) {
  if (!isSquare(arr)) return false;
  const n = arr.length;
  for (let i = 0; i < n; i++)
    for (let j = 0; j < n; j++)
      if (Math.abs(arr[i][j] - arr[j][i]) > 1e-10) return false;
  return true;
}

function isUpperTriangular(arr) {
  if (!isSquare(arr)) return false;
  for (let i = 1; i < arr.length; i++)
    for (let j = 0; j < i; j++)
      if (Math.abs(arr[i][j]) > 1e-10) return false;
  return true;
}

function isLowerTriangular(arr) {
  if (!isSquare(arr)) return false;
  for (let i = 0; i < arr.length; i++)
    for (let j = i + 1; j < arr[i].length; j++)
      if (Math.abs(arr[i][j]) > 1e-10) return false;
  return true;
}

function classificar_matriz(matrix) {
  const arr = (matrix && matrix.toArray) ? matrix.toArray() : matrix;
  const labels = [];
  if (isSquare(arr)) labels.push('square'); else labels.push('rectangular');
  if (isZero(arr)) labels.push('zero');
  if (isIdentity(arr)) labels.push('identity');
  if (isDiagonal(arr)) labels.push('diagonal');
  if (isSymmetric(arr)) labels.push('symmetric');
  if (isUpperTriangular(arr)) labels.push('upper_triangular');
  if (isLowerTriangular(arr)) labels.push('lower_triangular');
  return labels;
}

console.log('\n─── T3: Classificação ───');
console.log('Zero 3x4:', classificar_matriz(zeroMatrix));
console.log('Identidade 4x4:', classificar_matriz(identityMatrix));
console.log('Diagonal:', classificar_matriz(diagonalMatrix));
console.log('Triangular Superior:', classificar_matriz(upperTriangular));
console.log('Triangular Inferior:', classificar_matriz(lowerTriangular));
console.log('Simétrica:', classificar_matriz(symmetricMatrix));

// ─── T4: Erros de dimensões incompatíveis ─────────────────────────────────────

console.log('\n─── T4: Erros de dimensões ───');

try {
  const A = math.matrix([[1, 2, 3], [4, 5, 6]]);       // 2x3
  const B = math.matrix([[1, 2], [3, 4], [5, 6], [7, 8]]); // 4x2
  math.add(A, B);
} catch (e) {
  console.log('Erro ao somar 2x3 + 4x2:', e.message);
}

try {
  const A = math.matrix([[1, 2, 3], [4, 5, 6]]);   // 2x3
  const B = math.matrix([[1, 2], [3, 4], [5, 6]]); // 3x2 — ok
  const C = math.matrix([[1, 2, 3], [4, 5, 6]]);   // 2x3
  math.multiply(A, C); // 2x3 * 2x3 → dimensões internas incompatíveis
} catch (e) {
  console.log('Erro ao multiplicar 2x3 * 2x3:', e.message);
}
